k9mail/src/main/java/com/fsck/k9/activity/compose/AttachmentPresenter.java

```diff
@@ -0,0 +1,371 @@
+                public void onLoaderReset(Loader<Attachment> loader) {
+                    // nothing to do
+                }
+            };
+
+    private LoaderManager.LoaderCallbacks<Attachment> mAttachmentContentLoaderCallback =
+            new LoaderManager.LoaderCallbacks<Attachment>() {
+                @Override
+                public Loader<Attachment> onCreateLoader(int id, Bundle args) {
+                    Uri uri = args.getParcelable(LOADER_ARG_ATTACHMENT);
+                    return new AttachmentContentLoader(context, attachments.get(uri));
+                }
+
+                @Override
+                public void onLoadFinished(Loader<Attachment> loader, Attachment attachment) {
+                    int loaderId = loader.getId();
+                    loaderManager.destroyLoader(loaderId);
+
+                    if (!attachments.containsKey(attachment.uri)) {
+                        return;
+                    }
+
+                    if (attachment.state == Attachment.LoadingState.COMPLETE) {
+                        attachmentMvpView.updateAttachmentView(attachment);
+                        attachments.put(attachment.uri, attachment);
+                    } else {
+                        attachments.remove(attachment.uri);
+                        attachmentMvpView.removeAttachmentView(attachment);
+                    }
+
+                    postPerformStalledAction();
+                }
+
+                @Override
+                public void onLoaderReset(Loader<Attachment> loader) {
+                    // nothing to do
+                }
+            };
+
+    private void postPerformStalledAction() {
+        new Handler().post(new Runnable() {
+            @Override
+            public void run() {
+                performStalledAction();
+            }
+        });
+    }
+
+    void performStalledAction() {
+        attachmentMvpView.dismissWaitingForAttachmentDialog();
+
+        WaitingAction waitingFor = actionToPerformAfterWaiting;
+        actionToPerformAfterWaiting = WaitingAction.NONE;
+
+        switch (waitingFor) {
+            case SEND: {
+                attachmentMvpView.performSendAfterChecks();
+                break;
+            }
+            case SAVE: {
+                attachmentMvpView.performSaveAfterChecks();
+                break;
+            }
+        }
+    }
+
+    /**
+     * Add all attachments of an existing message as if they were added by hand.
+     *
+     * @param part
+     *         The message part to check for being an attachment. This method will recurse if it's
+     *         a multipart part.
+     * @param depth
+     *         The recursion depth. Currently unused.
+     *
+     * @return {@code true} if all attachments were able to be attached, {@code false} otherwise.
+     *
+     * @throws MessagingException
+     *          In case of an error
+     */
+    public boolean loadAttachments(Part part, int depth) throws MessagingException {
```

Comment:
this method is entirely unchanged, but I think it doesn't do what we want it to in all cases. The logic of what parts to use as attachments should also be extracted, but that's not in scope for this PR.